OAuth Token Workflow

inyono opened the pull request:
Adds the basic login / logout workflow with Hydra. Doesn't work in preview deployments since Hydra assumes that frontend runs on serlo-staging.dev resp. serlo.org (and we still redirect logged in users to the old system). To test locally:

- [Start the local dev environment for serlo.org](https://github.com/serlo/serlo.org#start).
- Run `yarn oauth:frontend` to create the OAuth client for the frontend in Hydra.

**TODO**
- [x] The Menu is only rendered on the client (as we discussed: we don't want to differentiate depending on auth state on server because of caching reasons). We might want to add an animation for that or something
- [x] Handle errors in `pages/api/auth` more gracefully @inyono
- [x] Login Workflow: after login, the page `src/pages/auth/login` is displayed for 5 seconds and then redirects to the previous page. We need to decide what to do here @elbotho (e.g. we could instead show a success notification and redirect directly)
- [x] Logout Workflow: after logout, the page `src/pages/auth/logout` is displayed for 5 seconds and then redirects to the previous page. We need to decide what to do here @elbotho (e.g. we could instead show a success notification and redirect directly)
- [ ] Logout should also clear the cookie in the legacy system @inyono ping https://github.com/serlo/serlo.org/issues/453
- [x] Use feature flag to disable this in production.

inyono committed:
oauth token workflow

Entkenntnis commented on src/pages/api/auth/callback.ts:
```diff
@@ -0,0 +1,42 @@
+      redirect_uri: 'http://localhost:3000/api/auth/callback',
+      scope,
+    })
+    const { token } = oauth2ClientCredentials.createToken(result)
```
Just wondering: why are we not using `oauth2AuthorizationCode.createToken()` here?

inyono replied:
Intended, the types are not up-to-date and this should be `clientCredentials` stuff

inyono committed:
fix callback

inyono committed:
fix callback

inyono commented on src/pages/api/auth/callback.ts:
```diff
@@ -0,0 +1,42 @@
+  if (csrf === req.cookies['auth-csrf']) {
+    const result = await oauth2AuthorizationCode.getToken({
+      code,
+      redirect_uri: 'http://localhost:3000/api/auth/callback',
```
@inyono this is wrong

inyono commented:
Result from the discussion with @elbotho:
- Login / logout workflow is most likely going to be a flasher
- @elbotho does the first iteration
- @inyono fixes remaining bugs in the auth workflow & tries to get this working on staging

elbotho committed:
add toast-notifications and fade in login menu part

- using `react-notify-toast`

elbotho committed:
add toast-notifications and fade in login menu part

- using `react-notify-toast`

elbotho commented:
Next router reloads page if we use it like this
``` tsx
void router.replace(props.referer.replace(document.location.origin, ''))
```
Instead to have a client side redirect we would need the same logic we use with Links:
``` tsx
<NextLink href="/[...slug]" as={decodeURIComponent(prettyLink)}>
```
(took me ages to figure out why the context was not persisting…)

So for now I'm using session storage to display the notification-toast, but's it's a bit of a workaround.
Good enough for now I'd say.
I keep the code in login and logout and see how https://github.com/serlo/frontend/issues/328 and https://github.com/serlo/frontend/issues/308 work out.

inyono committed:
oauth token workflow

# Conflicts:
#	.env.preview
#	.env.production
#	env.d.ts

# Conflicts:
#	.env.local
#	.env.production

inyono committed:
update checksums

inyono committed:
auth: fix callback url

inyono committed:
fix .env.production

inyono committed:
auth: fix callback, second attempt

inyono committed:
auth: fix callback, third attempt

inyono committed:
auth: seems like /api/auth/login is problematic, too

inyono commented:
Okay, this still doesn't work correctly since we have multiple redirect URIs and then OAuth requires a [sector validation uri](https://openid.net/specs/openid-connect-registration-1_0.html#SectorIdentifierValidation). Continuing with this another time -.-

elbotho committed:
Merge branch 'auth' of https://github.com/serlo/frontend into auth

elbotho committed:
fix merge, simpifly toast, cleanup

simplify for now:
- no provider for toast
- remove /auth/login and logout
- use hash to display notification

elbotho commented:
worked on the auth userflow again:

simplified it for now:
- no provider for toast
- redirect to referrer directly
- remove pages for /auth/login and logout
- use hash to display notification

feels a bit faster/nicer (no white page flash), code is quite clean and we can still expand it.

inyono commented:
Okay, going to look into that, thanks :)

inyono commented on src/components/navigation/header.tsx:
```diff
@@ -1,29 +1,35 @@
       </PaddedDiv>
       <SearchInput />
-      {isOpen && <MobileMenu links={menuData} />}
+      {isOpen && <MobileMenu links={links} />}
```
Mobile menu is missing auth stuff.

elbotho replied:
right, will there be useful functions for logged in mobile users? 
commenting I guess.

inyono replied:
Yes. With the new editor it should also be possible to make small edits on mobile. And any learner features that we introduce in the future.

elbotho replied:
My local auth setup is broken again, so I can't test thoroughly. 
But MobileMenu should now have the same functionality. It's a bit easier because we render the menu only when the user open it.

inyono commented on src/components/toast-notifications.tsx:
```diff
@@ -0,0 +1,29 @@
+import { useAuth } from '@/auth/use-auth'
+import { theme } from '@/theme'
+
+export function ToastNotifications() {
```
This should probably use `ToastContext` in the future, right?

elbotho replied:
it used to already, but I opted to use the simplest implementation for now and add in context etc. back in later when I understand the requirements of other notifications.

inyono replied:
Currently, `ToastContext` is not in use and you only handle auth notifications.

elbotho replied:
yes.

elbotho replied:
removed unused file.

inyono committed:
use history.replaceState instead of history.pushState

elbotho commented on package.json:
```diff
@@ -34,7 +34,7 @@
     "@tippyjs/react": "^4.0.0-alpha.0",
     "@types/nprogress": "^0.2.0",
-    "@types/react-notify-toast": "^0.5.2",
+    "@types/react-notify-toast": "^0.5.0",
```
@inyono did you downgrade this on purpose?

inyono replied:
Yes. My dependency workflow:
- Add the dependency with `^major.0.0` (resp. `^0.minor.0` for 0.x versions). If you need a specific feature (e.g. react hooks), specify the version that introduced the feature (i.e. `^16.8.0`).
- Use `yarn upgrade` to upgrade dependencies (instead of updating the version in `package.json`).

Reasoning: When some dependencies don't work together for some reason, we can look into `package.json` to see our "minimal" requirement for the dependency (and therefore know which dependencies we could downgrade).

So in this case, it's not an actual downgrade, since `yarn.lock` still contains the newest version. But it documents that we don't actually need `0.5.2`.

elbotho commented on src/components/toast-notifications.tsx:
```diff
@@ -16,8 +16,8 @@ export function ToastNotifications() {
-      //leave no trace
-      history.pushState(null, '', window.location.href.split('#')[0])
+      // Remove auth URI fragment
+      history.replaceState(null, '', window.location.href.split('#')[0])
```
thanks!

elbotho committed:
add auth menu to mobilemenu

elbotho committed:
remove toast context for now

inyono committed:
debug staging

inyono committed:
auth: use auth method client_secret_post

inyono committed:
auth: graceful error handling in /api/auth/callback

inyono committed:
auth: graceful error handling in /api/auth/login

inyono committed:
auth: graceful error handling in /api/auth/logout

inyono commented:
Auth now works correctly in staging (after setting `frontendDomain` to `frontend-git-auth.serlo.vercel.app`).

Entkenntnis committed:
Merge remote-tracking branch 'origin/staging' into auth

# Conflicts:
#	src/components/content/entity.tsx
#	src/components/navigation/menu.tsx
#	src/components/navigation/mobile-menu.tsx
#	src/data/menu.tsx

Entkenntnis committed:
add new login routes to exceptions

Entkenntnis committed:
don't use new auth in production

Entkenntnis committed:
rename

Entkenntnis committed:
replace server-side jsonwebtoken with client-side lib

Entkenntnis committed:
fix lint

Entkenntnis commented:
@inyono This is ready so far, auth is disabled for production (serlo.org), everything else looks fine.

inyono committed:
update package.json
